pkg/virt-handler/dmetrics-manager/dmetrics-manager.go

```diff
@@ -0,0 +1,144 @@
+package dmetrics_manager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"k8s.io/client-go/tools/cache"
+
+	v1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/downwardmetrics"
+	virtioserial "kubevirt.io/kubevirt/pkg/downwardmetrics/virtio-serial"
+	cmdclient "kubevirt.io/kubevirt/pkg/virt-handler/cmd-client"
+	"kubevirt.io/kubevirt/pkg/virt-handler/isolation"
+)
+
+func NewDownwardMetricsManager(nodeName string, virtSharedDir string, vmiInformer cache.SharedIndexInformer) (*DownwardMetricsManager, error) {
+	dmm := &DownwardMetricsManager{
+		done:          false,
+		nodeName:      nodeName,
+		virtSharedDir: virtSharedDir,
+		stopServer:    make(map[string]context.CancelFunc),
+	}
+
+	_, err := vmiInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			vmi, ok := obj.(*v1.VirtualMachineInstance)
+			if !ok {
+				log.Log.Errorf("v1.VirtualMachineInstance object failed conversion")
+				return
+			}
+			dmm.stopDMServer(vmi)
+		},
+		UpdateFunc: func(_, obj interface{}) {
+			vmi, ok := obj.(*v1.VirtualMachineInstance)
+			if !ok {
+				log.Log.Errorf("v1.VirtualMachineInstance object failed conversion")
+				return
+			}
+
+			err := dmm.startDMServer(vmi)
+			if err != nil {
+				log.Log.Reason(err).Errorf("failed to start the DownwardMetrics virtio-serial stopServer")
+			}
+		},
+	})
+
+	if err != nil {
+		return nil, err
+	}
+	return dmm, nil
+}
+
+// DownwardMetricsManager controls the lifetime of the DownwardMetrics servers.
+// Each server is tied to the lifetime of the VMI and DownwardMetricsManager itself.
+type DownwardMetricsManager struct {
+	lock          sync.Mutex
+	done          bool
+	nodeName      string
+	virtSharedDir string
+	stopServer    map[string]context.CancelFunc
+}
+
+// Run blocks until stopCh is closed. When done, it stops all remaining
+// running DownwardMetrics servers.
+func (m *DownwardMetricsManager) Run(stopCh chan struct{}) {
+	defer m.stop()
+	<-stopCh
+}
+
+func (m *DownwardMetricsManager) stop() {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	m.done = true
+
+	// Stop all DownwardMetrics servers
+	for domainName, stopServerFn := range m.stopServer {
+		stopServerFn()
+		delete(m.stopServer, domainName)
+	}
+}
+
+func (m *DownwardMetricsManager) stopDMServer(vmi *v1.VirtualMachineInstance) {
```

Comment:
Since we need to keep track of the already started servers in a map, we need to clean the vm entry in that map. I don't think it will hurt also to cancel the server context